internal/pkg/agent/application/upgrade/step_download.go

```diff
@@ -46,32 +50,63 @@ func (u *Upgrader) downloadArtifact(ctx context.Context, version, sourceURI stri
 
 	pgpBytes = u.appendFallbackPGP(version, pgpBytes)
 
+	parsedVersion, err := agtversion.ParseVersion(version)
+	if err != nil {
+		return "", fmt.Errorf("error parsing version %q: %w", version, err)
+	}
+
 	// do not update source config
 	settings := *u.settings
+	var downloaderFunc downloader
+	var factory downloaderFactory
+	var verifier download.Verifier
 	if sourceURI != "" {
 		if strings.HasPrefix(sourceURI, "file://") {
 			// update the DropPath so the fs.Downloader can download from this
 			// path instead of looking into the installed downloads directory
 			settings.DropPath = strings.TrimPrefix(sourceURI, "file://")
+
+			// use specific function that doesn't perform retries on download as its
+			// local and no retry should be performed
+			downloaderFunc = u.downloadOnce
+
+			// set specific downloader, local file just uses the fs.NewDownloader
+			// no fallback is allowed because it was requested that this specific source be used
+			factory = func(ver *agtversion.ParsedSemVer, l *logger.Logger, config *artifact.Config, d *details.Details) (download.Downloader, error) {
```

Comment:
you can even move this to fs and have `factory = fs.NewDownloaderFactory` to make it a bit more readable

Reply:
i didn't do that as the factory interface is private, takes arguments that the download module doesn't really need to import as its not used